### ISSUE TITLE
Integration with vimspector

### ISSUE THREAD

**Issue by danilbykov** [scalameta/coc-metals#80]
**Describe the feature**

This vim plugin (puremourning/vimspector) implements DebugAdapterProtocol which was recently added in metals. At this moment it supports several languages (C, Java, Python, javascript and so on). Are you interested in integration between coc-metals and vimspector?

**Potential ways to implement?**

In order to start debug session in vimspector, user should define file `.vimspector.json`. This file describes how to start Debug Server, path to program, environment variables and so on. In scala we don't need these options because metals can start Debug Server for us. So we don't need whole file `.vimspector.json`. Vimspector should just connect to this server and start talking. I'm going to open Pull Request for vimspector and implement this.

So general flow is
1. Using CodeLens from LSP we start DebugAdapter in Metals
2. Metals returns port where DebugAdapter is started
3. coc-metals calls vimspector and provides this port.

I was able to debug simple test application using this schema.

**Additional context**

vimspector is enough buggy plugin but hopefully they will be able to polish it.

**Comment by ckipp01** [comment]
Hey @danilbykov! Thanks for opening the request. I've been doing some thinking about this lately, so I'm glad you reached out.

The ability to not need a `.vimspector.json` file would be great. One of the things that was keeping me from really diving into vimspector was that I was a bit skeptical of how much setup was needed, and how it would/could integrate with `coc-metals`. So, a couple question and comments.

## Regarding your general flow outline
1. How do you envision the CodeLens working? In Neovim we could use virtual text for example, but that would have to probably appear at the end of a line? Would there just be a key binding to trigger the process when on that line? With Vim and the lack of virtual text, how do you see that working?
2/3. So if I understand it correctly, `coc-metals` would simply provides the "setup" or location of the necessary port to vimspector, and then it would take over. There really wouldn't be much need for a lot of chance in the `coc-metals` codebase correct?

Thanks again for starting the conversation on this.

**Comment by danilbykov** [comment]
Hi Chris,

1. Do you know that coc-nvim already supports CodeLens in neovim? You just need to set `codeLens.enable` in CocConfig. It is supported based on virtual text. You also need to define some hotkey, for example,
```
nnoremap <leader>cl :<C-u>call       CocActionAsync('codeLensAction')<CR>
```

Yes, it is not supported in Vim. Maybe we can implement this for vim without virtual text. We need to highlight somehow line with CodeLens, maybe underline this text.

2. Yes, it is small change in coc-metals. I have branch with required changes (not ready for PR):
https://github.com/scalameta/coc-metals/compare/master...danilbykov:dap

**Comment by ckipp01** [comment]
Hey @danilbykov,

Yea, I'm aware that they support lenses, but wasn't fully aware of the `codeLensAction`. Thanks. I do think it's important that if we do it, we have a way for vim users to use the feature as well, _if possible_.

Regarding the minimal changes, that looks fantastic. I'd say yea, keep working on it, and feel free to send in a pr when you're ready for it to be looked at! 🎉

**Comment by puremourning** [comment]
> vimspector is enough buggy plugin but hopefully they will be able to polish it.

What bugs have you found? Could you please raise them as issues on GitHub as opposed to off hand unrelated comments?

**Comment by danilbykov** [comment]
@puremourning, I just quoted this from FAQ from README.md.
> Q: Does it work? A: Yeah, sort of. It's incredibly buggy and unpolished.
You probably should edit this section if it is no longer true.

I saw couple issues but I think they are specific to neovim and you already know about them, for example, it looks like Watches window does not in neovim.

**Comment by puremourning** [comment]
>  it looks like Watches window does not work in neovim.

It works. Read [the documentation](https://github.com/puremourning/vimspector#neovim-differences) - yes `:VimspectorWatch` works.

> You probably should edit this section if it is no longer true.

I may be able to edit that in future, but I won't be able to un-make your claim here.

**Comment by danilbykov** [comment]
@ckipp01,

@puremourning proposes to use a bit different approach than I described initially. In order integrate coc-metals and vimspector an user needs to create file `/path/to/vimspector/configurations/<os>/scala/cocmetals.json` with following content:
```
{
  "configurations": {
    "Default": {
      "adapter": {
        "port": "${port}",
        "variables": {
        }
      },
      "configuration": {
        "name": "default",
        "request": "launch"
      },
      "breakpoints": {
        "exception": {
          "caught": "N",
          "uncaught": "N"
        }
      }
    }
  }
}
```
It must be done only once. Vimspector can connect to DebugAdapter Server with this file. I think this is good approach because it still doesn't require too much configuring from user.

But I have another question. At this moment I can run or debug any application but I can not specify environment parameters and main class arguments. We need to find some way to define these parameters. Then coc-metals can pass these variables to metals inside command:
```
Params: {
  "command": "debug-adapter-start",
  "arguments": [
    {
      "targets": [
        {
          "uri": "file:/home/dbykov/reps/test/experiments/?id\u003droot"
        }
      ],
      "dataKind": "scala-main-class",
      "data": {
        "class": "monix.HotObservable",
        "arguments": [
        ],
        "jvmOptions": [
        ]
      }
    }
  ]
}
```
We can define some file `debug.json` in root of project with following content:
```
[
{
  "mainClass": "com.github.full.path.to.class",
  "arguments": [
    "arguments",
    "for",
    "main",
    "methods"
  ],
  "jvmOptions": [
    "jvm",
    "options"
  ]
}, {
  ...
}
]
```
Then coc-metals can analyze this file and pass predefined arguments (if any) to Metals. What do you think?

BTW: There is similar [feature-request](https://github.com/scalameta/metals-feature-requests/issues/63) for VS code.

**Comment by ckipp01** [comment]
Hi @danilbykov, I think a bit of manual setup in the beginning is ok. I'd much rather get it working with a bit of setup, and then start to smooth out the process with potentially the file being automatically created by command, which shouldn't be too hard to do. As long as everything is documented well, I think we're good.

Regarding the `debug.json`, from my initial understanding, that seems like a good idea. However, I think we can also split that up into a separate pr. I'd rather get the integration working, and then tackle this. I'm also going to ping @tgodzik just to make sure this all sounds sane. When we enter DAP stuff I'm pretty unfamiliar with it.

**Comment by tgodzik** [comment]
> Hi @danilbykov, I think a bit of manual setup in the beginning is ok. I'd much rather get it working with a bit of setup, and then start to smooth out the process with potentially the file being automatically created by command, which shouldn't be too hard to do. As long as everything is documented well, I think we're good.
> 
> Regarding the `debug.json`, from my initial understanding, that seems like a good idea. However, I think we can also split that up into a separate pr. I'd rather get the integration working, and then tackle this. I'm also going to ping @tgodzik just to make sure this all sounds sane. When we enter DAP stuff I'm pretty unfamiliar with it.

We could reuse the `launch.json` that is used in VSCode - we could do it then inside of the common library.

https://code.visualstudio.com/docs/editor/debugging#_launch-configurations

**Comment by danilbykov** [comment]
https://github.com/scalameta/coc-metals/pull/83

**Comment by puremourning** [comment]
> Regarding the debug.json, from my initial understanding, that seems like a good idea. However, I think we can also split that up into a separate pr. I'd rather get the integration working, and then tackle this. I'm also going to ping @tgodzik just to make sure this all sounds sane. When we enter DAP stuff I'm pretty unfamiliar with it.

`debug.json` ? `launch.json` that's just replicating what `.vimspector.json` is for, but less powerful. I don't get why you would want that. Vimspector will send the config in the initialise request just like vscode would with `launch.json`.

Why would you invent a new file that's outside of the DAP ?

**Comment by danilbykov** [comment]
In our schema Metals runs DebugAdapter server. As I know only coc-metals (or metals) can connect to Metals. Environment parameters must be specified when DebugAdapter server is started. I don't understand how vimspector can help this.

**Comment by puremourning** [comment]
@danilbykov other things connect to metals, like YCM, but that's hardly relevant.

The way DAP works is:

* <something> starts the debug adapter (in this case, metals)
* Initialisation exchange, with protocol configuration
* Launch request (or attach request) with *adapter specific* configuration (which is what .vimspector.json and launch.json provide)

You're suggesting some third-config file which would live outside of the DAP, which I don't see the value in. 

The java debug adapter works the same way as metals, where the debug adapter is embedded in the language server and opens a TCP port, but it works like a normal debug adapter for the sake of configuration and launch/attach requests.

**Comment by puremourning** [comment]
On the other hand, if configuration is required for launching the debug adapter itself, then vimspector provides a powerful mechanism to do that too (as does vscode, though you have to write an extension in typescript or something). If that's the case, then what value is metals adding by launching the adapter for you? Is it adding more context or something ?

**Comment by ckipp01** [comment]
Just jumping in to drop a quick note. Thanks for the pr in #83 , @danilbykov, and thanks for following along and for the valuable input @puremourning .

Regarding the implementation for this, I'll need to dive in a bit and test some things now that we have a pr. Especially regarding the `launch.json/debug.json` file, if we don't actually need it, and all of that information can just be passed in via the `.vimspector.json` file, then that'd be preferable.

Thanks again for the contribution and discussion.

**Comment by danilbykov** [comment]
@puremourning ,

Here is log from language server. Parameters for application are passed already here (arguments.data.{arguments&jvmOptions})
```
[Trace - 08:42:25 PM] Received request 'workspace/executeCommand - (9)'
Params: {
  "command": "debug-adapter-start",
  "arguments": [
    {
      "targets": [
        {
          "uri": "file:/home/dbykov/reps/test/experiments/?id\u003droot"
        }
      ],
      "dataKind": "scala-main-class",
      "data": {
        "class": "monix.HotObservable",
        "arguments": [
          "param0",
          "param1",
          "param2"
        ],
        "jvmOptions": [
          "-Dtest.opt1\u003d123",
          "-Dtest.opt2\u003d234",
          "-Xmx1G"
        ]
      }
    }
  ]
}

[Trace - 08:42:25 PM] Sending response 'workspace/executeCommand - (9)'. Processing request took 111ms
Result: {
  "name": "monix.HotObservable",
  "uri": "tcp://0.0.0.0:36653"
}
```
and related logs from debug adapter
```
[Trace][11:00:14 PM] Received request:
{
  "type": "request",
  "seq": 0,
  "command": "initialize",
  "arguments": {
    "adapterID": "adapter",
    "clientID": "vimspector",
    "clientName": "vimspector",
    "linesStartAt1": true,
    "columnsStartAt1": true,
    "locale": "en_GB",
    "pathFormat": "path",
    "supportsVariableType": true,
    "supportsVariablePaging": false,
    "supportsRunInTerminalRequest": true
  }
}
[Trace][11:00:14 PM] Sent response:
{
  "type": "response",
  "seq": 1,
  "request_seq": 0,
  "command": "initialize",
  "success": true,
  "body": {
    "supportsConfigurationDoneRequest": true,
    "supportsHitConditionalBreakpoints": true,
    "supportsConditionalBreakpoints": true,
    "supportsEvaluateForHovers": true,
    "supportsCompletionsRequest": true,
    "supportsRestartFrame": true,
    "supportsSetVariable": true,
    "supportsRestartRequest": false,
    "supportTerminateDebuggee": true,
    "supportsDelayedStackTraceLoading": false,
    "supportsLogPoints": true,
    "supportsExceptionInfoRequest": true,
    "exceptionBreakpointFilters": [
      {
        "label": "Uncaught Exceptions",
        "filter": "uncaught"
      },
      {
        "label": "Caught Exceptions",
        "filter": "caught"
      }
    ]
  }
}
[Trace][11:00:14 PM] Received request:
{
  "type": "request",
  "seq": 1,
  "command": "launch",
  "arguments": {
    "name": "default",
    "request": "launch"
  }
}
[Trace][11:00:14 PM] Sent notification:
{
  "type": "event",
  "seq": 2,
  "event": "initialized",
  "body": {
    "type": "initialized"
  }
}
[Trace][11:00:14 PM] Sent response:
{
  "type": "response",
  "seq": 2,
  "request_seq": 1,
  "command": "launch",
  "success": true
}
[Trace][11:00:14 PM] Received request:
{
  "type": "request",
  "seq": 2,
  "command": "setExceptionBreakpoints",
  "arguments": {
    "filters": []
  }
}
[Trace][11:00:14 PM] Sent response:
{
  "type": "response",
  "seq": 3,
  "request_seq": 2,
  "command": "setExceptionBreakpoints",
  "success": true
}
[Trace][11:00:14 PM] Received request:
{
  "type": "request",
  "seq": 3,
  "command": "configurationDone"
}
[Trace][11:00:14 PM] Sent response:
{
  "type": "response",
  "seq": 4,
  "request_seq": 3,
  "command": "configurationDone",
  "success": true
}
[Trace][11:00:14 PM] Received request:
{
  "type": "request",
  "seq": 4,
  "command": "threads"
}
 ```

>  if configuration is required for launching the debug adapter itself

So I think that this is our case.


> If that's the case, then what value is metals adding by launching the adapter for you?

Metals provide CodeLens and vim already can highlight "debuggable" files (applications or tests). So I can press one hotkey and start debugging.

**Comment by puremourning** [comment]
I don't see how CodeLens and "starting a debug adapter process" are related.

**Comment by danilbykov** [comment]
Metals sends CodeLens with command which starts debug adapter.
```
[Trace - 08:42:23 PM] Sending response 'textDocument/codeLens - (8)'. Processing request took 76ms
Result: [
  {
    "range": {
      "start": {
        "line": 10,
        "character": 7
      },
      "end": {
        "line": 10,
        "character": 20
      }
    },
    "command": {
      "title": "debug",
      "command": "metals-debug-session-start",
      "arguments": [
        {
          "targets": [
            {
              "uri": "file:/home/dbykov/reps/test/experiments/?id\u003droot"
            }
          ],
          "dataKind": "scala-main-class",
          "data": {
            "class": "monix.HotObservable",
            "arguments": [],
            "jvmOptions": []
          }
        }
      ]
    }
  }
]
```

It is late here. I need to sleep. I will be able to continue tomorrow. Thanks for your help.

**Comment by puremourning** [comment]
Oh I see, in that case you don't need a config file because the debug adapter arguments are sent from the language server to the client and back on executeCommand request.

**Comment by ckipp01** [comment]
https://github.com/scalameta/metals-vscode/pull/281
